### ISSUE TITLE
Robust handling of the new chart form

### ISSUE THREAD

**Issue by lfborjas** [lfborjas/freenatalchart.xyz#9]
See the spectre reference for forms to display all of these errors in an accessible manner:

https://picturepan2.github.io/spectre/elements/forms.html

For interacting with the Algolia and Google APIs, look into the [`req`](https://github.com/mrkkrp/req) package -- I've used `wreq` before but... it's too much man, the lenses, the wonky types...!

Also, don't forget to add a [`noscript`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/noscript) in case scripting is disabled, warning about possible degraded geolocation experience.

## API references:

* Algolia REST API: https://community.algolia.com/places/api-clients.html#alternative-json-answer-format (there's also a free geonames API to fall back to: http://www.geonames.org/export/geonames-search.html, example `curl` [here](https://gist.github.com/lfborjas/c8deef5e3fe535a787aa4998c2a272bf))
* Geonames timezone lookup: https://www.geonames.org/export/web-services.html#timezone
* NOTE: maybe the location fallback should happen through geonames.org, too? It seems a little bit slower than Algolia, and surely less accurate, but since it's likely that lat/lng may be unavailable because of an error with Algolia, this seems like a safer fallback?

For the [weird responses](https://www.geonames.org/export/webservice-exception.html) from geonames: https://williamyaoh.com/posts/2019-10-19-a-cheatsheet-to-json-handling.html#parsing-weird-json-formats

### Rabbit hole: the timezone fallback, with a smidgen of the lookup fallback too

tl;dr: start with geonames, because it's free and has pretty good rates (though presumably slow and inaccurate?) and look into upgrading to askGeo when we've deployed, if we can get approved for free usage!

Check out:

* https://data.iana.org/time-zones/tz-link.html
* Which leads to: https://stackoverflow.com/questions/16086962/how-to-get-a-time-zone-from-a-location-using-latitude-and-longitude-coordinates/16086964
* And also: https://github.com/evansiroky/timezone-boundary-builder which could be used via: https://github.com/BertoldVdb/ZoneDetect in conjunction with https://hackage.haskell.org/package/timezone-olson -- however, this homebrewing sounds tough (both in terms of development and the data I'd have to bundle, in particular the boundaries,) but possible if we _really_ want to go offline. 

Runners up, as suggested by [that IANA page](https://data.iana.org/time-zones/tz-link.html):

> Free access via a network API, if you register a key, is provided by the GeoNames Timezone web service, the Google Maps Time Zone API, and the Time Zone Database & API. Commercial network API access is provided by AskGeo and GeoGarage.

[Geonames](http://www.geonames.org/export/web-services.html#timezone) -- one would pay [40 euros a year](http://www.geonames.org/commercial-webservices.html) for the the commercial service (compare with the... thousands one would pay using Google,)

The free service has [a pretty decent limit, too](https://www.geonames.org/export/#ws):

> 20'000 credits daily limit per application (identified by the parameter 'username'), the hourly limit is 1000 credits. A credit is a web service request hit for most services. An exception is thrown when the limit is exceeded.

There's also [TimezoneDB](https://timezonedb.com/references/get-time-zone) which is [$5/month for premium](https://timezonedb.com/premium) but which has a 1 req/second limit. 

Then of course there's [Google Places](https://developers.google.com/maps/documentation/timezone/overview), but [the billing is quite high](https://developers.google.com/maps/documentation/timezone/usage-and-billing): for the same 20,000 daily requests allowed by the free geonames service, one would pay $100 dollars!

And lastly, there's [AskGeo](https://askgeo.com), which is apparently [faster and more accurate than geonames](https://stackoverflow.com/questions/41504/timezone-lookup-from-latitude-longitude/5519523#5519523), but the pricing is quite high -- but definitely not as high as google! -- ($29/mo, 2,000,000 credits.) -- however, they seem to offer [free access for non-commercial projects](https://askgeo.com/#pricing), so we may look into applying if geonames is a pain in the butt? One caveat here is that we'd need a different geolocation service, as askgeo doesn't have that capability. 

## Location

If `lat` and `lng` aren't present, it could be due to JS being disabled or Algolia being down. The server should try to:

* Take the (required) `location` argument and call Algolia on its own. 
* If algolia returns no results (quota reached, or just wonky place,) display the form with an error.
* If algolia returns one result, use that lat and lng.
* [optiona/future?] If algolia returns multiple results, display the form with those as a radio group (whose value is the combined lat/lng?) 

## Timezone

We try to determine timezones with the aid of Google. If Google fails, we should either also fail, or ask for the timezone in the form? Leaning towards the former 😅, but the latter is viable since we're simply rendering server-side.

## Date parsing

Since we're obtaining the date parts separately, if they're present it should be trivial to compose a `UTCTime`. But I brought in the [`validation-selective`](https://hackage.haskell.org/package/validation-selective) package anyway, for here be dragons.
